File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/OfficeHoursLogic.kt
package com.lextures.android.core.lms

import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.time.format.FormatStyle

/** Office-hours slot formatting, filtering, and planner calendar mapping (M7.3). */
object OfficeHoursLogic {
    const val CALENDAR_EVENT_TITLE = "Office Hours"
    const val VIRTUAL_LABEL = "Virtual"

    fun isMyBooking(slot: AppointmentSlot): Boolean =
        slot.status == "booked" && !slot.studentId.isNullOrBlank()

    fun upcomingAvailableSlots(slots: List<AppointmentSlot>, now: Instant = Instant.now()): List<AppointmentSlot> =
        slots
            .filter { it.status == "available" }
            .filter { slot ->
                val start = LmsDates.parse(slot.slotStart) ?: return@filter false
                !start.isBefore(now)
            }
            .sortedBy { it.slotStart }

    fun myBookedSlots(slots: List<AppointmentSlot>, now: Instant = Instant.now()): List<AppointmentSlot> =
        slots
            .filter(::isMyBooking)
            .filter { slot ->
                val start = LmsDates.parse(slot.slotStart) ?: return@filter false
                !start.isBefore(now)
            }
            .sortedBy { it.slotStart }

    fun formatSlotTime(slot: AppointmentSlot): String {
        val start = LmsDates.parse(slot.slotStart) ?: return slot.slotStart
        val end = LmsDates.parse(slot.slotEnd)
        val zone = ZoneId.systemDefault()
        val startText = DateTimeFormatter.ofLocalizedDateTime(FormatStyle.MEDIUM, FormatStyle.SHORT)
            .withZone(zone)
            .format(start)
        if (end == null) return startText
        val endText = DateTimeFormatter.ofLocalizedTime(FormatStyle.SHORT)
            .withZone(zone)
            .format(end)
        return "$startText – $endText"
    }

    fun locationLabel(window: AvailabilityWindow?): String? {
        if (window == null) return null
        val location = window.location?.trim().orEmpty()
        if (location.isEmpty()) {
            return if (window.isVirtual) VIRTUAL_LABEL else null
        }
        return if (window.isVirtual) {
            "$VIRTUAL_LABEL · $location"
        } else {
            location
        }
    }

    fun windowMap(windows: List<AvailabilityWindow>): Map<String, AvailabilityWindow> =
        windows.associateBy { it.id }

    fun calendarEvents(
        courseCode: String,
        courseTitle: String,
        slots: List<AppointmentSlot>,
        windows: List<AvailabilityWindow>,
    ): List<PlannerCalendarEvent> =
        myBookedSlots(slots).mapNotNull { slot ->
            val start = LmsDates.parse(slot.slotStart) ?: return@mapNotNull null
            val end = LmsDates.parse(slot.slotEnd)
            PlannerCalendarEvent(
                id = "office-hours:${slot.id}",
                title = CALENDAR_EVENT_TITLE,
                courseCode = courseCode,
                courseTitle = courseTitle,
                startsAt = start,
                endsAt = end,
                allDay = false,
                kind = PlannerCalendarEventKind.OfficeHours,
                officeHoursSlotId = slot.id,
                meetingId = slot.meetingId,
            )
        }

    fun collectCalendarEvents(
        studentCourses: List<CourseSummary>,
        availabilityByCourseCode: Map<String, OfficeHoursAvailability>,
    ): List<PlannerCalendarEvent> =
        studentCourses.flatMap { course ->
            if (!course.isOfficeHoursEnabled) return@flatMap emptyList()
            val availability = availabilityByCourseCode[course.courseCode] ?: return@flatMap emptyList()
            calendarEvents(course.courseCode, course.displayTitle, availability.slots, availability.windows)
        }
}

val CourseSummary.isOfficeHoursEnabled: Boolean
    get() = officeHoursEnabled == true

File: clients/ios/Lextures/Core/LMS/OfficeHoursLogic.swift
import Foundation

/// Office-hours slot formatting, filtering, and planner calendar mapping (M7.3).
enum OfficeHoursLogic {
    static func isMyBooking(_ slot: AppointmentSlot) -> Bool {
        slot.status == "booked" && slot.studentId != nil
    }

    static func upcomingAvailableSlots(_ slots: [AppointmentSlot], now: Date = Date()) -> [AppointmentSlot] {
        slots
            .filter { $0.status == "available" }
            .filter { guard let start = LMSDates.parse($0.slotStart) else { return false }; return start >= now }
            .sorted { ($0.slotStart) < ($1.slotStart) }
    }

    static func myBookedSlots(_ slots: [AppointmentSlot], now: Date = Date()) -> [AppointmentSlot] {
        slots
            .filter(isMyBooking)
            .filter { guard let start = LMSDates.parse($0.slotStart) else { return false }; return start >= now }
            .sorted { ($0.slotStart) < ($1.slotStart) }
    }

    static func formatSlotTime(_ slot: AppointmentSlot) -> String {
        guard LMSDates.parse(slot.slotStart) != nil else { return slot.slotStart }
        let startText = DateFormatting.formatDateTime(slot.slotStart)
        guard LMSDates.parse(slot.slotEnd) != nil else { return startText }
        let endText = endTimeOnly(slot.slotEnd)
        return "\(startText) – \(endText)"
    }

    private static func endTimeOnly(_ raw: String) -> String {
        guard let date = DateFormatting.parse(raw) else { return raw }
        let formatter = DateFormatter()
        formatter.locale = LocalePreferences.effectiveLocaleValue()
        formatter.timeZone = .current
        formatter.dateStyle = .none
        formatter.timeStyle = .short
        return formatter.string(from: date)
    }

    static func locationLabel(window: AvailabilityWindow?) -> String? {
        guard let window else { return nil }
        let location = window.location?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        guard !location.isEmpty else {
            return window.isVirtual ? L.text("mobile.officeHours.virtual") : nil
        }
        return window.isVirtual ? "\(L.text("mobile.officeHours.virtual")) · \(location)" : location
    }

    static func windowMap(_ windows: [AvailabilityWindow]) -> [String: AvailabilityWindow] {
        Dictionary(uniqueKeysWithValues: windows.map { ($0.id, $0) })
    }

    static func calendarEvents(
        courseCode: String,
        courseTitle: String,
        slots: [AppointmentSlot],
        windows: [AvailabilityWindow]
    ) -> [PlannerCalendarEvent] {
        let lookup = windowMap(windows)
        return myBookedSlots(slots).compactMap { slot in
            guard let start = LMSDates.parse(slot.slotStart) else { return nil }
            let end = LMSDates.parse(slot.slotEnd)
            return PlannerCalendarEvent(
                id: "office-hours:\(slot.id)",
                title: L.text("mobile.officeHours.calendarTitle"),
                courseCode: courseCode,
                courseTitle: courseTitle,
                startsAt: start,
                endsAt: end,
                allDay: false,
                kind: .officeHours,
                structureKind: nil,
                structureItemId: nil,
                notebookPageId: nil,
                officeHoursSlotId: slot.id,
                meetingId: slot.meetingId
            )
        }
    }

    static func collectCalendarEvents(
        studentCourses: [CourseSummary],
        availabilityByCourseCode: [String: OfficeHoursAvailability]
    ) -> [PlannerCalendarEvent] {
        studentCourses.flatMap { course in
            guard course.isOfficeHoursEnabled,
                  let availability = availabilityByCourseCode[course.courseCode] else { return [PlannerCalendarEvent]() }
            return calendarEvents(
                courseCode: course.courseCode,
                courseTitle: course.displayTitle,
                slots: availability.slots,
                windows: availability.windows
            )
        }
    }
}
